### PR TITLE
Broadcast plugin written and tested

### LINKED ISSUE #159: Broadcast checks written and tested

### PR DESCRIPTION
**Pull request by nwalsh1995**
Resolves #159 

There is a test java file which contains calls to all the methods this plugin looks for. The plugin basically looks for calls to these methods:

```
    "sendBroadcast",
    "sendBroadcastAsUser",
    "sendOrderedBroadcast",
    "sendOrderedBroadcastAsUser",
    "sendStickyBroadcast",
    "sendStickyBroadcastAsUser",
    "sendStickyOrderedBroadcast",
    "sendStickyOrderedBroadcastAsUser",
```

### CODE REVIEW

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        self.current_file = None
+        self.manifest_xml = None
+        self.below_min_sdk_21 = False
+
+    def run(self, files, apk_constants=None):
+        try:
+            self.below_min_sdk_21 = apk_constants["min_sdk"] < 21
+        except (KeyError, TypeError):
+            for decompiled_file in files:
+                if decompiled_file.lower().endswith("androidmanifest.xml"):
```
Comment (dhaugh) [inline qark/plugins/broadcast/send_broadcast_receiver_permission.py]: Couldn't we just filter for androidmanifest.xml? Would it start with something else?

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                    self.below_min_sdk_21 = get_min_sdk(manifest_xml=minidom.parse(decompiled_file)) < 21
+                else:
+                    self.below_min_sdk_21 = False
+
+        java_files = (decompiled_file for decompiled_file in files if decompiled_file.lower().endswith(".java"))
+        for java_file in java_files:
+            with open(java_file, "r") as java_file_to_read:
```
Comment (dhaugh) [inline qark/plugins/broadcast/send_broadcast_receiver_permission.py]: this should be wrapped in a try/except

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py (hunk 1 above)
Comment (nwalsh-lnk) [reply qark/plugins/broadcast/send_broadcast_receiver_permission.py]: `decompiled_file` is going to be the absolute path for these files. The thing we can do is to see if it ends with `/AndroidManifest.xml` or `\AndroidManifest.xml`, those approaches are a bit hacky but I don't know of another way. I'm open to ideas though.

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            self.current_file = java_file
+            for method_invocation in parsed_tree.filter(MethodInvocation):
+                self._check_method_invocation(method_invocation, parsed_tree.imports)
+
+    def _check_method_invocation(self, method_invocation, imports):
+        """
+        Checks the type of method that is being called and the amount of arguments and adds it to the issues list if it
+        results in a warning or vulnerability.
+
+        :param javalang.tree.MethodInvocation method_invocation: method being called
```
Comment (dhaugh) [inline qark/plugins/broadcast/send_broadcast_receiver_permission.py]: Isn't this a tuple of node_path, Node?

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        :param list imports: list of imports
+        """
+        method_name = method_invocation[1].member
+        num_arguments = len(method_invocation[1].arguments)
+        if method_invocation[1].member in BROADCAST_METHODS:
```
Comment (dhaugh) [inline qark/plugins/broadcast/send_broadcast_receiver_permission.py]: 1) use method_name, not method_invocation[1].member
2) check for the inverse (not in BROADCAST_METHODS) and return immediately - saves a level of nesting

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py (hunk 1 above)
Comment (dhaugh) [reply qark/plugins/broadcast/send_broadcast_receiver_permission.py]: import ntpath
ntpath.basename(<path>)

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        """
+        Checks the type of method that is being called and the amount of arguments and adds it to the issues list if it
+        results in a warning or vulnerability.
+
+        :param javalang.tree.MethodInvocation method_invocation: method being called
+        :param list imports: list of imports
+        """
+        method_name = method_invocation.member
+        num_arguments = len(method_invocation.arguments)
+        if method_invocation.member in BROADCAST_METHODS:
```
Comment (dhaugh) [inline qark/plugins/broadcast/send_broadcast_receiver_permission.py]: Check for not in BROADCAST_METHODS, and return immediately if not present. Saves a level of indentation

File: tests/test_java_files/send_broadcast_receiver_permission.java [hunk 6]
```diff
@@ -0,0 +1,418 @@
+/*
```
Comment (dhaugh) [inline tests/test_java_files/send_broadcast_receiver_permission.java]: Would it be possible to break this up into individual java files and run the test on each one of them? It's non obvious how many errors are in this file

File: tests/test_scanner/test_scanner.py [hunk 7]
```diff
@@ -16,6 +16,11 @@ def test_run_manifest_checks(scanner):
     assert 0 < len(scanner.issues)
 
 
+def test_run_broadcast_checks(scanner):
+    scanner._run_broadcast_checks()
+    assert 0 < len(scanner.issues)
```
Comment (dhaugh) [inline tests/test_scanner/test_scanner.py]: This should be an exact number, not a non-zero number

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py (hunk 1 above)
Comment (nwalsh-lnk) [reply qark/plugins/broadcast/send_broadcast_receiver_permission.py]: I've looked through the docs and think we can just use `os.sep` instead.

File: tests/test_java_files/send_broadcast_receiver_permission.java (hunk 6 above)
Comment (nwalsh-lnk) [reply tests/test_java_files/send_broadcast_receiver_permission.java]: I've modified the java file so its a lot easier to read.

File: qark/plugins/broadcast/send_broadcast_receiver_permission.py (hunk 5 above)
Comment (nwalsh-lnk) [reply qark/plugins/broadcast/send_broadcast_receiver_permission.py]: It saves a level of indention but makes logic harder to follow similar to what we discussed yesterday. Going to leave it for now

**Pull request merged**